clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -2000,6 +2003,55 @@ protected void processCommonConsumerAction(
     }
   }
 
+  /**
+   * This function checks, for a store, if its persisted offset is greater than the end offset of its
+   * corresponding Kafka topic (indicating an offset rewind event, thus a potential data loss in Kafka) and increases
+   * related sensor counter values.
+   */
+  private void reportStoreVersionTopicOffsetRewindMetrics(PartitionConsumptionState pcs) {
+    long offset = pcs.getLatestProcessedLocalVersionTopicOffset();
+    if (offset == OffsetRecord.LOWEST_OFFSET) {
+      return;
+    }
+    // Proceed if persisted OffsetRecord exists and has meaningful content.
+    long endOffset = getKafkaTopicPartitionEndOffSet(localKafkaServer, kafkaVersionTopic, pcs.getPartition());
```

Comment:
I have a minor performance concern: will this affect metrics reporting performance when we hit Kafka issue such as topic not exist or slowness error? 

(I remember we have encountered slow metrics retrieval and it turned out that it is kafka offset retrieval slowness..but maybe I could be wrong here..)

Reply:
Jialin, thank you for your review and raising this for discussion. To better understand your concerns, do you worry about this function could end up reporting too many metics? or do you worry that getKafkaTopicPartitionEndOffSet could be very slow, thus making the latency of calling this method very high?

For point #1, this metric check happens when server subscribes to a new topic partition and only reports when a rewind is detected, which in normal case, shouldn't be very often.

For point #2, getKafkaTopicPartitionEndOffSet first tries to get the end offset from the `KafkaConsumerService` and thus `TopicPartitionsOffsetsTracker` who knows the offset kind of free when consuming a Kafka record. If it is not available there, then it asks for `CachedKafkaMetadataGetter` who queries Kafka servers, but also caches Kafka responses for 60 seconds for reuse.

Does it make sense, or maybe you referred to something else?

Reply:
Thanks for this detailed replies, I think I was talking about 2, and seems like this is covered well. Thank you! Please feel free to mark it as resolved.